Additions to Chatbot Pattern (#446)
carbon-design-system/carbon-website: src/pages/experimental/chatbot/flows.mdx

---
title: Chatbots
description: Use a chatbot when a user would benefit from requesting information on-demand with natural language. They should receive a useful and trustworthy response related to their question or task.
tabs: ['Overview', 'Usage', 'Flows', 'Content']
---

<AnchorLinks>

> <AnchorLink>Launching a Bot</AnchorLink>

Reviewer: ```suggestion
<AnchorLink>Launching a bot</AnchorLink>
```

> <AnchorLink>Multiple Threads</AnchorLink>

Reviewer: ```suggestion
<AnchorLink>Multiple threads</AnchorLink>
```

> <AnchorLink>Expanding Cards</AnchorLink>

Reviewer: ```suggestion
<AnchorLink>Expanding cards</AnchorLink>
```

> <AnchorLink>Error States</AnchorLink>

Reviewer: ```suggestion
<AnchorLink>Error states</AnchorLink>
```

</AnchorLinks>

> ## Launching a Bot

Reviewer: ```suggestion
## Launching a bot
```

Bots should be launched from a button positioned at the bottom right of the screen, or from a button embedded within the UI. See Carbon Chatbot Add-on Design Kit for the appropriate variants.

> ![Launching a Bot](images/chatbot-5.png)

Reviewer: ```suggestion
![Launching a bot](images/chatbot-5.png)
```

> ## Multiple Threads

Reviewer: ```suggestion
## Multiple threads
```

This flow demonstrates connecting the user to a human agent when the bot cannot complete the request or the user would prefer interacting with a person.

> ### User Initiates Change

Reviewer: ```suggestion
### User initiates change
```

Sometimes, a user indicates that they would like to chat with a human. Other times, this process can be suggested if a bot is repeatedly unsuccessful in answering a user’s question.

> ![User Initiates Change](images/chatbot-12.png)

Reviewer: ```suggestion
![User initiates change](images/chatbot-12.png)
```

> ### Notify the User

Reviewer: ```suggestion
### Notify the user
```

Use a toast notification bar to indicate when the user is in a new conversation context, or that a new user has joined the the chat. Call out any significant changes with in-conversation status messages.

![Notify the User](images/chatbot-13.png)

> ### Multiple Users

Reviewer: ```suggestion
### Multiple users
```

Combining two or more human agents into the same thread is a common practice. Where necessary, call out significant changes in who the user is addressing (particularly if they have different purposes)

> ![Multiple Users](images/chatbot-14.png)

Reviewer: ```suggestion
![Multiple users](images/chatbot-14.png)
```

> ### Incoming Messages

Reviewer: ```suggestion
### Incoming messages
```

For new messages appearing in another thread, be sure to use dots above the Threads icon to alert the user of new messages. Then, within the menu, indicate the amount of new messages on the specific thread. Visually distinguish threads with new messages.

> ![Incoming Messages](images/chatbot-15.png)

Reviewer: ```suggestion
![Incoming messages](images/chatbot-15.png)
```

> ## Expanding Cards

Reviewer: ```suggestion
## Expanding cards
```

Where needed, use expanding cards to progressively disclose content. Use a full page view to show a very large amount of content to the user.

> ### Default State

Reviewer: ```suggestion
### Default state
```

Call out critical information in the default state. Provide a trigger to expand the card

> ![Default State](images/chatbot-16.png)

Reviewer: ```suggestion
![Default state](images/chatbot-16.png)
```

> ### Expanded State

Reviewer: ```suggestion
### Expanded state
```

Provide more detailed information in the expanded state. Only if absolutely necessary, provide a way to access the full view of the content

![Expanded State](images/chatbot-17.png)

> ### Full View

Reviewer: ```suggestion
### Full view
```

The full view is a generic container that can be used to display long-form content that would otherwise crowd a conversation thread.

> ![Full View](images/chatbot-18.png)

Reviewer: ```suggestion
![Full view](images/chatbot-18.png)
```

> ## Error States (Experimental)

Reviewer: ```suggestion
## Error states (Experimental)
```

Be sure to indicate to a user when their message has not been sent due to a connectivity or server problem. When in a connected state, temporarily show the success notification, and then remove it along with any message error indicators.

> ![Chatbot Error Notification](images/chatbot-6.png)

Reviewer: ```suggestion
![Chatbot error notification](images/chatbot-6.png)
```

---
title: Chatbots
description: Use a chatbot when a user would benefit from requesting information on-demand with natural language. They should receive a useful and trustworthy response related to their question or task.
tabs: ['Overview', 'Usage', 'Flows', 'Content']
---

### Depending on the complexity of the use case and user needs, chatbots may require other interactions that supplement the conversational experience.

> This page documents some common chatbot scenarios in IBM products and their recommended best practices. Flows still under consideration / user testing are marked with "Experimental".

Reviewer: ```suggestion
<InlineNotification>

**Note:**This page documents some common chatbot scenarios in IBM products and their recommended best practices. Flows still under consideration are marked with "Experimental".

</InlineNotification>
```

<AnchorLink>Multiple threads</AnchorLink>
<AnchorLink>Expanding cards</AnchorLink>
<AnchorLink>Error states</AnchorLink>

</AnchorLinks>

## Launching a bot

Bots should be launched from a button positioned at the bottom right of the screen, or from a button embedded within the UI. See Carbon Chatbot Add-on Design Kit for the appropriate variants.

![Launching a bot](images/chatbot-5.png)

## Multiple threads (Experimental)

When a bot cannot complete a request, the bot detects frustration, or the user would simply prefer interacting with a person, it becomes necessary to connect the user to a human agent.

### User initiates change

Sometimes, a user indicates that they would like to chat with a human. Other times, this process can be suggested if a bot is repeatedly unsuccessful in answering a user’s question.

![User initiates change](images/chatbot-12.png)

### Notify the user

Use a toast notification bar to indicate when the user is in a new conversation context, or that a new user has joined the the chat. Call out any significant changes with in-conversation status messages.

![Notify the user](images/chatbot-13.png)

### Multiple users

> Combining two or more human agents into the same thread is a common practice. Where necessary, call out significant changes in who the user is addressing (particularly if they have different purposes)

Reviewer 2: ```suggestion
Combining two or more human agents into the same thread is a common practice. Where necessary, call out significant changes in who the user is addressing (particularly if they have different purposes).
```

![Multiple users](images/chatbot-14.png)

### Incoming messages

For new messages appearing in another thread, be sure to use dots above the Threads icon to alert the user of new messages. Then, within the menu, indicate the amount of new messages on the specific thread. Visually distinguish threads with new messages.

![Incoming messages](images/chatbot-15.png)

## Expanding cards (Experimental)

Where needed, use expanding cards to progressively disclose content. Use a full page view to show a very large amount of content to the user.

### Default state

> Call out critical information in the default state. Provide a trigger to expand the card

Reviewer 2: ```suggestion
Call out critical information in the default state. Provide a trigger to expand the card.
```

![Default state](images/chatbot-16.png)

### Expanded state

> Provide more detailed information in the expanded state. Only if absolutely necessary, provide a way to access the full view of the content

Reviewer 2: ```suggestion
Provide more detailed information in the expanded state. Only if absolutely necessary, provide a way to access the full view of the content.
```
